File: 0560-subarray-sum-equals-k/main.py
from collections import defaultdict

class Solution:
    def subarraySum(self, nums, k):
        prefix_count = defaultdict(int)
        prefix_count[0] = 1

        prefix_sum = 0
        result = 0

        for num in nums:
            prefix_sum += num
            need = prefix_sum - k

            if need in prefix_count:
                result += prefix_count[need]

            prefix_count[prefix_sum]+=1
        return result
    
if __name__ == "__main__":
    sol = Solution()
    print(sol.subarraySum([1, 1, 1], 2))   # 2
    print(sol.subarraySum([1, 2, 3], 3))   # 2
    print(sol.subarraySum([1, 2, 1, 2, 1], 3))  # 4

File: 0560-subarray-sum-equals-k/main.cpp
#include <unordered_map>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> prefixCount;
        
        prefixCount[0] = 1;

        int prefixSum = 0;
        int result = 0;

        for(int num: nums){
            prefixSum += num;
            int need = prefixSum-k;
            if(prefixCount.find(need)!= prefixCount.end()){
                result += prefixCount[need];
            }
            prefixCount[prefixSum]++;
        }

        return result;
    }
};